src/main/java/org/opensearch/security/action/apitokens/ApiToken.java

```diff
@@ -0,0 +1,239 @@
+
+        @Override
+        public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+            builder.startObject();
+            builder.array(INDEX_PATTERN_FIELD, indexPatterns.toArray(new String[0]));
+            builder.array(ALLOWED_ACTIONS_FIELD, allowedActions.toArray(new String[0]));
+            builder.endObject();
+            return builder;
+        }
+    }
+
+    /**
+     * Class represents an API token.
+     * Expected class structure
+     * {
+     *   name: "token_name",
+     *   jti: "encrypted_token",
+     *   creation_time: 1234567890,
+     *   cluster_permissions: ["cluster_permission1", "cluster_permission2"],
+     *   index_permissions: [
+     *     {
+     *       index_pattern: ["index_pattern1", "index_pattern2"],
+     *       allowed_actions: ["allowed_action1", "allowed_action2"]
+     *     }
+     *   ],
+     *   expiration: 1234567890
+     * }
+     */
+    public static ApiToken fromXContent(XContentParser parser) throws IOException {
+        String name = null;
+        String jti = null;
+        List<String> clusterPermissions = new ArrayList<>();
+        List<IndexPermission> indexPermissions = new ArrayList<>();
+        Instant creationTime = null;
+
+        XContentParser.Token token;
+        String currentFieldName = null;
+
+        while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
+            if (token == XContentParser.Token.FIELD_NAME) {
+                currentFieldName = parser.currentName();
+            } else if (token.isValue()) {
+                switch (currentFieldName) {
+                    case NAME_FIELD:
+                        name = parser.text();
+                        break;
+                    case JTI_FIELD:
+                        jti = parser.text();
+                        break;
+                    case CREATION_TIME_FIELD:
+                        creationTime = Instant.ofEpochMilli(parser.longValue());
+                        break;
+                }
+            } else if (token == XContentParser.Token.START_ARRAY) {
+                switch (currentFieldName) {
+                    case CLUSTER_PERMISSIONS_FIELD:
+                        while (parser.nextToken() != XContentParser.Token.END_ARRAY) {
+                            clusterPermissions.add(parser.text());
+                        }
+                        break;
+                    case INDEX_PERMISSIONS_FIELD:
+                        while (parser.nextToken() != XContentParser.Token.END_ARRAY) {
+                            if (parser.currentToken() == XContentParser.Token.START_OBJECT) {
+                                indexPermissions.add(parseIndexPermission(parser));
+                            }
+                        }
+                        break;
+                }
+            }
+        }
+
+        if (name == null) {
+            throw new IllegalArgumentException(NAME_FIELD + " is required");
```

Comment:
can this be empty? if not you can leverage isNullOrEmpty supplied by core

Reply:
Another option would be to create a method for this validation checks:
```java
public void validateFields(String... fields) throws IllegalArgumentException {
   for (String field : fields) { 
   
   // switch block to validate against the field variables declared at the start of the class
   }
}
```

Reply:
You can also remove it there are validations in place to prevent null or Empty values from being fed into the index document.